Fixed #36281 -- Used async-safe write in ASGIHandler.read_body().

zebra0345 opened the pull request:
#### Trac ticket number
ticket-36281

#### Branch description
This change ensures the "read_body" method
In "ASGIHandler" uses "sync_to_async" with "thread_sensitive=False" for file writing.
This makes the method fully async-safe and avoids potential issues when handling large or streamed HTTP request bodies.

Additionally, a comprehensive test method was added to validate the behavior across multiple cases:
- Multiple chunks
- Single chunk
- Empty chunks
- Disconnection mid-body

#### Checklist
- [x] This PR targets the `main` branch.
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

carltongibson commented on django/core/handlers/asgi.py:
```diff
@@ -255,6 +255,10 @@ async def read_body(self, receive):
         )
+
+        # async_safe version
+        async_write = sync_to_async(body_file.write, thread_sensitive=False)
```
This incurs significant overhead for request bodies smaller than `FILE_UPLOAD_MAX_MEMORY_SIZE`, which in normal cases we can expect to be most (or all) of them.

zebra0345 replied:
Thanks for your advice

I've attempted to implement the branching logic as advised, using "getattr" to safely check for rollover.

Not 100% sure if I captured it exactly right, but the updated tests pass and seem to behave correctly.

Really appreciate your feedback and time

carltongibson reviewed (commented):
I think we need to be more sophisticated than the initial suggestion here. 

For comparison, Starlette checks whether a SpooledTemporaryFile is still in memory or not before running the write in a threadpool in a similar case (that of an uploaded file): 

https://github.com/encode/starlette/blob/eb6cebbeecbbc44f87acb310330b37813d06598d/starlette/datastructures.py#L437-L444

carltongibson commented:
Thanks for the update @zebra0345. I need to think through the test cases. 

Just thinking out loud: I'd still prefer us to somehow *allow* folks to use `aiofiles` here. The minimum (I'd guess) would be to make the class used for `body_file` into a class attribute and move the `write` block into a method, so that one could subclass `ASGIRequest` to use the alternate async file type. This might be better addressed separately.

zebra0345 commented:
Thank you for the suggestion!

Although this is an easy picking ticket, I'd like to align the implementation more closely with Django's design philosophy.  
I’ll revise the patch to follow your proposed structure — moving the `body_file` class to a class attribute and extracting the write logic into a separate method — and submit an updated PR accordingly.

If I understood correctly, the idea is:
1. Refactor the write block into a method (e.g., `write_body_chunk()`), and  
2. Move the file creation logic to a class attribute like `body_file_class`,  
   so that users can subclass and override it more easily.

Thanks again for your guidance — I really appreciate your time and thoughtful feedback!

zebra0345 commented:
> @zebra0345 Wait on implementing that! It might be better as a follow-up (if at all) ([aiofiles is just performing the same trick under the hood anyway...](https://github.com/Tinche/aiofiles/blob/f76b7589ef2dcb7017c6508147db03b41d3f7747/src/aiofiles/tempfile/temptypes.py#L37-L39))

Got it — thanks for the clarification!

I’ll stick with the current implementation for this PR, then.  
Happy to explore the structural change as a follow-up later if needed.

Appreciate the continued guidance!

carltongibson commented on django/core/handlers/asgi.py:
```diff
@@ -255,6 +255,7 @@ async def read_body(self, receive):
         body_file = tempfile.SpooledTemporaryFile(
             max_size=settings.FILE_UPLOAD_MAX_MEMORY_SIZE, mode="w+b"
         )
+
```
Please revert this unrelated change to keep the diff minimal.

carltongibson commented on django/core/handlers/asgi.py:
```diff
@@ -263,7 +264,16 @@ async def read_body(self, receive):
             # Add a body chunk from the message, if provided.
             if "body" in message:
-                body_file.write(message["body"])
+                body = message["body"]
+
+                # Check if the file has rolled over to disk
+                if getattr(body_file, "_rolled", False):
+                    # Use async-safe write
+                    await sync_to_async(body_file.write, thread_sensitive=False)(body)
+                else:
+                    # Still in memory: safe to write synchronously.
+                    body_file.write(body)
+
```
I think I'd adjust this slightly: 

```suggestion
                # Perform IO in another thread if the temp file has rolled over
                # to disk.
                on_disk = getattr(body_file, "_rolled", False):
                if on_disk:
                    async_write = sync_to_async(
                        body_file.write, 
                        thread_sensitive=False,
                    )
                    await async_write(message["body"])
                else:
                    body_file.write(message["body"])
```

zebra0345 replied:
Thank you for the helpful review!

I've updated the code according to your suggestion — simplified the write logic and reverted the unrelated change to keep the diff minimal.

Please let me know if there's anything else I should adjust. I really appreciate your time and thoughtful guidance throughout this process!

carltongibson reviewed (commented):
Hi @zebra0345. 

This looks good — just a couple of initial comments. 

I will look closely at the tests shortly.

carltongibson commented on django/core/handlers/asgi.py:
```diff
@@ -263,7 +263,20 @@ async def read_body(self, receive):
+                        body_file.write,
+                        thread_sensitive=False,
+                    )
+                    # Use async_write
+                    await async_write(message["body"])
+                else:
+                    # Still in memory: safe to write synchronously.
+                    body_file.write(message["body"])
```
Given the block comment and the code, I don't think we need the additional line comments here.

carltongibson commented on tests/asgi/tests.py:
```diff
@@ -2,14 +2,17 @@
 import sys
 import threading
 import time
+from contextlib import (
+    nullcontext,  # Used in test_read_body_comprehensive_cases (ticket #36281)
+)
```
Please single-line this and remove the comment. 

```suggestion
from contextlib import nullcontext
```

zebra0345 replied:
Hi @carltongibson , thank you for the detailed review!

You're absolutely right — the current test doesn't directly assert the thread context of the write() call.

I'll revise the test to patch the write method and explicitly check which thread it's being called from (in-memory vs. on-disk). Once I’ve got that in place, I’ll update the PR accordingly.

Appreciate your guidance as always!

carltongibson reviewed (commented):
Hi @zebra0345 — so I've left a few more comments, please look at those. 

The extra tests are OK. I think they're in part duplicating what we already have. 
(I find the tests a little difficult to follow with the subtest usage, but maybe that's OK. I'd lean to a more specific case, showing just what's at stake, see below.)

Main issue is that they're not testing the issue in hand. If I revert `asgi.py` to before 2ec2424a54148e74ce003cef2cd08b7298663ecb (your initial commit) none of the new test cases fail. 

I think we need to assert the thread that the write call is executed in. In the in-memory case, it should be the same thread as the current event loop. For the rolled case it should be a different thread, for the thread pool executor. I guess we'll need to patch `write` to make that check accessible. 

I think if we can demonstrate that, we've got an OK looking solution here.

I'm going to mark the ticket as Needs Tests on Trac. Please unflag that when you've added the tests and I think it'll be ready for a review from one of the Fellows. 

Thanks for the work here.

zebra0345 commented:
Thank you again for your continued feedback and guidance throughout this!

I tried to provide a more precise test to verify whether write() is truly offloaded to a threadpool when rollover occurs.
To that end, I compared the thread name of each write call — and confirmed that when the file rolls over, the write is indeed performed in a separate thread (as expected).

Attached screenshots:

One shows the debug code I used to log the thread info.
![debug_code](https://github.com/user-attachments/assets/a55cdc4e-cacd-41ff-a9c8-f03afc6e1b28)

The other is the result, showing ThreadPoolExecutor-XX being used for disk-backed writes.
![result](https://github.com/user-attachments/assets/b4a65472-0737-4bfc-89f0-33f8f19be3af)

I’ve also removed the extra line comments you mentioned in the ASGI handler, and kept the test focused and as clear as possible.
It was really insightful to learn more about Django’s memory rollover behavior, and I appreciate your suggestions which helped me structure this properly.

If you could double-check whether the test logic now captures the intended behavior correctly, I'd really appreciate it!

Thanks again!

zebra0345 commented:
Hi @carltongibson,

Just following up on this PR — I’ve updated the patch based on your feedback and ensured the test demonstrates the fix.  
All checks are passing, and the Trac ticket status has been updated as well.

If it’s ready for another review (by you or someone else), I’d really appreciate the next step.

Thanks again for all the help so far!

carltongibson commented:
@zebra0345 Great thanks. (This PR crossed my mind the other day…) I will take another look this week.

zebra0345 commented:
Thanks a lot,
I really appreciate you taking the time. I'll wait for your review patiently!

carltongibson commented on tests/asgi/tests.py:
```diff
@@ -659,3 +661,58 @@ async def test_streaming_disconnect(self):
         with self.assertRaises(asyncio.TimeoutError):
             await communicator.receive_output(timeout=0.2)
+
+    # test read_body() for ticket #36281
```
Please remove this comment.

carltongibson reviewed (approved):
This looks good to me @zebra0345 — thanks for the effort 👍

carltongibson commented on tests/asgi/tests.py:
```diff
@@ -659,3 +661,58 @@ async def test_streaming_disconnect(self):
+                    with patch.object(f, "write", side_effect=write_wrapper):
+                        await handler.read_body(receive_rolled)
+
+        self.assertTrue(any(t != loop_thread for t in called_threads))
```
```suggestion
        # The second write should have rolled over to disk.
        self.assertTrue(any(t != loop_thread for t in called_threads))
```

zebra0345 commented:
Hi @carltongibson,

I've removed the comment as suggested and also added the clarifying comment above the thread check.

All checks have passed again.

Please let me know if there's anything else I should adjust — thank you for the continued guidance!

zebra0345 commented:
Thanks @carltongibson and @felixxm for the thoughtful reviews and support throughout this process!
Really appreciated the guidance, learned a lot through this contribution. Looking forward to contributing more in the future!
